web/docs/write-a-package.md

```diff
@@ -0,0 +1,327 @@
+      | chart bar
+```
+
+## Make the package configurable
+
+After we have illustrated how the package works with examples, let's step back
+for a moment and assess how customizable the package should be:
+
+- Is there anything that might differ from user to user?
+- Do they have to bring their API key for the package to work?
+- Are timeouts highly subjective and specific to the local environment?
+
+For all places where it's difficult to offer a one-size-fits-all assumption, we
+want the user to make decision on how to proceed. These customization points are
+called *inputs*, and the correspondingly named section in the package definition
+specifies them.
+
+In our case, we hard-coded the refresh interval that updates the SSLBL lookup
+table to exactly 1 hour. Maybe other users want to update just once a day? Or
+more often? This is a typical example where policy is user-specific and where we
+can turn an assumption into a decision. Let's define the input:
+
+```yaml title="package.yaml"
+inputs:
+  refresh-interval:
+    name: Time between context updates
+    description: |
+      The duration between updates that fetch the SSLBL database via the API.
+    default: 1 hour
+```
+
+By specifying a default value, we also make it easy for users to skip the
+decision making. In other words, defaults make a configuration knob optional.
+
+After we've defined our configuration knobs, we now go over the pipeline
+definitions and replace the hard-coded constants with a placeholder:
+
+```yaml title="package.yaml"
+pipelines:
+  update-context:
+    name: Update SSLBL Context
+    description: |
+      A pipeline that periodically refreshes the SSLBL lookup table.
+    definition: |
+      // tql2
+      every {{ inputs.refresh-interval }} {
+        load_http "from https://sslbl.abuse.ch/blacklist/sslblacklist.csv"
+        read_csv comments=true, header="timestamp,SHA1,reason"
+        legacy "context update sslbl --key=SHA1"
+      }
+    restart-on-error: 1 hour
+```
+
+Note how we simply replaced `1h` with `{{ inputs.refresh-interval }}`.
+
+## Test your package
+
+When you think you're done, it's time to validate that things work as you
+expect. This means effectively trying to install, configure and use it.
+
+:::note Testing Framework
+We currently don't offer a native testing framework for packages where you can
+provide tests and baselines along the package definition. But we love the idea,
+and if you do as well, please swing by our [Community Discord](/discord) and
+discuss it with us.
+:::
+
+[Installing a package](installation/install-a-package.md) given a `package.yaml`
+file is easiest with the [`package_add`](operators/package.md) operator, since
+a package is just data:
+
+```tql
+load_file "/path/to/package.yaml"
+read_yaml
+package_add
+```
+
+This fails with the following error:
+
+```
+error: failed to add package
+ = note: with error: !! unspecified: named argument `header` does not exist
+ = note: failed to add package
```

Comment:
This error threw me off. I think we need to improve the message.

Reply:
@IyeOnline is `header="a,b,c"` the right syntax after #4527?

Reply:
The input is substituted with an empty string in your case, which is the default value. The syntax for the header you assumed is correct.